pkg/source/ttnv3/config.go

```diff
@@ -15,121 +14,197 @@ import (
 
 var logger *zap.SugaredLogger
 
-type config struct {
-	appID string
-
+type serverConfig struct {
+	applicationServerGRPCAddress,
 	identityServerGRPCAddress,
 	joinServerGRPCAddress,
-	applicationServerGRPCAddress,
 	networkServerGRPCAddress string
+}
+
+func (c *serverConfig) SetEmptyFields(address string) {
+	if a := &c.applicationServerGRPCAddress; *a == "" {
+		a = &address
+	}
+	if a := &c.identityServerGRPCAddress; *a == "" {
+		a = &address
+	}
+	if a := &c.joinServerGRPCAddress; *a == "" {
+		a = &address
+	}
+	if a := &c.networkServerGRPCAddress; *a == "" {
+		a = &address
+	}
+}
+
+func (c *serverConfig) AnyFieldEmpty() error {
+	if c.applicationServerGRPCAddress == "" {
+		return errNoApplicationServerGRPCAddress.New()
+	}
+	if c.identityServerGRPCAddress == "" {
+		return errNoIdentityServerGRPCAddress.New()
+	}
+	if c.joinServerGRPCAddress == "" {
+		return errNoJoinServerGRPCAddress.New()
+	}
+	if c.networkServerGRPCAddress == "" {
+		return errNoNetworkServerGRPCAddress.New()
+	}
+	return nil
+}
+
+type Config struct {
+	appID string
+
+	serverConfig serverConfig
 
 	deleteSourceDevice,
 	dryRun,
 	noSession bool
 }
 
+var config Config
```

Comment:
So if my understanding is correct here, you have added this global state in order to work around the fact that you cannot pass state between `flagSet()` and `NewSource`.

Historically (in other sources) it is indeed the case that the idea was that these two are decoupled - you created the flags with some kind of `flagSet()`, then read the flags from the flag set in `NewSource`. But please note that nothings stops you from actually having state pass between the two.

You can have `flagSet` also return a `*Config`, and then the `NewSource` function can be built using this `*Config`:
```go
func createNewSource(config *Config) func(context.Context, *pflag.FlagSet) (source.Source, error) {
	return func(ctx context.Context, _ *pflag.FlagSet) (source.Source, error) {
		return Source{
			ctx:    ctx,
			config: config,
		}, nil
	}
}
```

I know that `getConfig` also does some parsing but you can consider moving that to this `newSource` factory function.

You can consider even dropping the `*pflag.FlagSet` from the `NewSource` function interface - this is a bit more work but I encourage it.